en/11_Compute_Shader.md

```diff
@@ -0,0 +1,605 @@
+
+    vkUpdateDescriptorSets(device, 3, descriptorWrites.data(), 0, nullptr);
+}
+```
+
+Remember that we also have to request the descriptor types for the SSBOs from our descriptor pool:
+
+```c++
+std::array<VkDescriptorPoolSize, 2> poolSizes{};
+...
+
+poolSizes[1].type = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER;
+poolSizes[1].descriptorCount = static_cast<uint32_t>(MAX_FRAMES_IN_FLIGHT) * 2;
+```
+
+We need to double the number of `VK_DESCRIPTOR_TYPE_STORAGE_BUFFER` types requested from the pool by two because our sets reference the SSBOs of the last and current frame.
+
+## Compute pipelines
+
+As compute is not a part of the graphics pipeline, we can't use `vkCreateGraphicsPipelines` to attach the compute shader to it. Instead we need to create a dedicated compute pipeline `vkCreateComputePipelines` for running our compute commands. Since a compute pipeline does not touch any of the rasterization state, it has a lot less state than a graphics pipeline:
+
+```c++
+VkComputePipelineCreateInfo pipelineInfo{};
+pipelineInfo.sType = VK_STRUCTURE_TYPE_COMPUTE_PIPELINE_CREATE_INFO;
+pipelineInfo.layout = computePipelineLayout;
+pipelineInfo.stage = computeShaderStageInfo;
+
+if (vkCreateComputePipelines(device, VK_NULL_HANDLE, 1, &pipelineInfo, nullptr, &computePipeline) != VK_SUCCESS) {
+    throw std::runtime_error("failed to create compute pipeline!");
+}
+```
+
+The setup is a lot simpler, as we only require one shader stage and a pipeline layout. The pipeline layout works the same as with the graphics pipeline:
+
+```c++
+VkPipelineLayoutCreateInfo pipelineLayoutInfo{};
+pipelineLayoutInfo.sType = VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO;
+pipelineLayoutInfo.setLayoutCount = 1;
+pipelineLayoutInfo.pSetLayouts = &computeDescriptorSetLayout;
+
+if (vkCreatePipelineLayout(device, &pipelineLayoutInfo, nullptr, &computePipelineLayout) != VK_SUCCESS) {
+    throw std::runtime_error("failed to create compute pipeline layout!");
+}
+```
+
+## Compute space
+
+Before we get into how a compute shader works and how we submit compute workloads to the GPU, we need to talk about two important compute concepts: **work groups** and **invocations**. They define an abstract execution model for how compute workloads are processed by the compute hardware of the GPU in three dimensions (x,y and z).
+
+**Work groups** define how the compute workloads are formed and processed by the the compute hardware of the GPU. You can think of them as work items the GPU has to work through. Work group dimensions are set by the application at command buffer time using a dispatch command.
+
+And each work group then is a collection of **invocations** that execute the same compute shader. Invocations can potentially run in parallel and their dimensions are set in the compute shader. Invocations within a single workgroup have access to shared memory.
+
+This image shows the relation between these two in three dimensions:
+
+![](/images/compute_space.svg)
```

Comment:
Maybe this image should have `Dispatch [3,3,3]`, indicating that its a 3x3x3 dispatch? What I'm thinking is that there is no obvious indication of why there is a 3 wide cube being shown.

Similar thought applies to the invocations.

Reply:
So something like this?

![image](https://user-images.githubusercontent.com/10283127/206767821-49c03b6b-8077-4501-807b-8fc365978ad3.png)